votes/0150-pr-issue-association.md

```diff
@@ -0,0 +1,51 @@
+# A Vote for PR Issue Association
+
+* Discussion at https://github.com/tikv/community/pull/150
+* Tracking issue https://github.com/tikv/community/issues/149
+
+## Proposal
+
+### Motivation
+
+To measure software evolution, we focus on features implemented, bugs fixed, and general enhancement. Issues document these items, while a pull request focuses on the change set and whether we implement it correctly.
+
+Normally, an issue documents a work item, and a pull request resolves it (despite cherry-picks). However, the relation between pull requests and issues is complicated in TiDB projects.
+
+* 1 issue with 0 PRs: question, bug vanished without a reason, etc. 
+* 1 issue with 1 PR: normal case. 
+* 1 issue with n PRs: resolve a bug multiple times, or resolve a complex issue in steps.
+* n issues with 1 PR: trade off resolving things as a whole (due to code coupling, or we can resolve multiple bugs at the common parent level), duplicated issues. 
+* **0 issue with 1 PR**: coding without a ticket ([example](https://github.com/tikv/tikv/pull/10999)). Even if the description is detailed or even if a design document is present, pull requests focus on the change set. It is about the implementation, not the question; if it is, we always mix discussion because the change set has a primary position. 
+
+### Solution
+
+Previous discussion happened on the [developer discussion forum](https://internals.tidb.io/t/topic/409).
+
+I propose to check that a pull request must be associated with issue(s).
+
+In details:
+
+* A pull request MUST be associated with issue(s) in commit messages. That is, the change set by `git commit` with a message mention the issue(s) associated, by `ref #{issue-number}` or `close[sd]?|resolve[sd]?|fix(e[sd])? #{issue-number}`.
```

Comment:
Commits in a PR are usually squashed into one. How the ensure that the commit message is still there after getting merged?

Reply:
AFAIK, the PR title is always in the commit message. And I also remember that you've discussed marking the issue number in the title. Why is this approach not considered?

Reply:
> AFAIK, the PR title is always in the commit message. And I also remember that you've discussed marking the issue number in the title. Why is this approach not considered?

It is listed as alternative [here](https://github.com/tikv/community/pull/150/files#diff-9cd0dc74f6dc2697f6b983446bcdaf892b9414113f3efe37990c7f48f4fd6155R39-R41). I tend to the approach in commit message and we're going to propose one solution. So there is primary and secondary.

Reply:
> Commits in a PR are usually squashed into one. How the ensure that the commit message is still there after getting merged?

@Mini256 can give more information on technical part. IIUC `tide` will squash all commit messages and we don't omit it once it is checked.

Reply:
@andylokandy i'd oppose to making the issue number in title. this will turn the commit message summary into a mess like `*: fix something [#12345] (#23456)`, and even worse in cherry pick `*: fix something [#12345] (#23456) (#23460)`.

Reply:
@andylokandy @tisonkun 

When tide merges PRs in the squash way, it will use the commit message provided by GitHub by default (because [the commit message body](https://github.com/ti-community-infra/configs/blob/02b3d4f51c98fdf7fbbaaa643e746fbec165bba7/prow/config/config.yaml#L1111) configured under the tikv organization is nil). The default commit message will merge multiple commit messages in the PR into one list. And it will ignore the commit generated during the merge-base.

Refer: [merge-message-for-a-squash-merge](https://docs.github.com/en/github/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#merge-message-for-a-squash-merge)